builder/builder.go

```diff
@@ -0,0 +1,386 @@
+	if b.err != nil {
+		return b
+	}
+
+	if b.locals == nil && b.inputs == nil && b.selectors == nil &&
+		b.variants == nil && b.patterns == nil { // simple start text
+		s = textEscape(s, true)
+	} else {
+		s = textEscape(s, false)
+	}
+
+	if len(b.variants) == 0 {
+		b.patterns = append(b.patterns, s)
+	} else {
+		b.variants[len(b.variants)-1].pattern = append(b.variants[len(b.variants)-1].pattern, s)
+	}
+
+	return b
+}
+
+func (b *Builder) Local(v string, expr *Expression) *Builder {
+	if b.err != nil {
+		return b
+	}
+
+	if len(b.patterns) > 0 {
+		b.err = fmt.Errorf("complex message cannot be added after simple message")
+		return b
+	}
+
+	expr.spacing = b.spacing
+
+	b.locals = append(b.locals, local{variable: variable(v), expr: expr})
+
+	return b
+}
+
+func (b *Builder) Input(expr *Expression) *Builder {
+	if b.err != nil {
+		return b
+	}
+
+	if len(b.patterns) > 0 {
+		b.err = fmt.Errorf("complex message cannot be added after simple message")
+		return b
+	}
+
+	expr.spacing = b.spacing
+
+	b.inputs = append(b.inputs, *expr)
+
+	return b
+}
+
+func (b *Builder) Expr(e *Expression) *Builder {
+	if b.err != nil {
+		return b
+	}
+
+	e.spacing = b.spacing
+
+	if len(b.variants) == 0 {
+		b.patterns = append(b.patterns, e)
+	} else {
+		b.variants[len(b.variants)-1].pattern = append(b.variants[len(b.variants)-1].pattern, e)
+	}
+
+	return b
+}
+
+func (b *Builder) Match(selectors ...*Expression) *Builder {
+	if b.err != nil {
+		return b
+	}
+
+	if len(b.patterns) > 0 {
+		b.err = fmt.Errorf("complex message cannot be added after simple message")
+		return b
+	}
+
+	if len(selectors) == 0 {
+		b.err = fmt.Errorf("match MUST have at least one selector")
+		return b
+	}
+
+	for i := range selectors {
+		b.selectors = append(b.selectors, *selectors[i])
+	}
+
+	return b
+}
+
+func (b *Builder) Key(keys ...any) *Builder {
+	if b.err != nil {
+		return b
+	}
+
+	if len(keys) != len(b.selectors) {
+		b.err = fmt.Errorf("number of keys in each variant MUST match the number of selectors in the matcher")
+		return b
+	}
+
+	b.variants = append(b.variants, variant{keys: keys})
+
+	return b
+}
+
+type variant struct {
+	keys, pattern []any
+}
+
+func (v *variant) String() string {
+	var s string
+
+	for i := range v.keys {
+		switch v := v.keys[i].(type) {
+		case string:
+			s += v + " "
+		case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
+			s += fmt.Sprintf("%d ", v)
+		case float32, float64:
+			s += fmt.Sprintf("%f ", v)
+		default:
+			panic(fmt.Sprintf("unsupported key type: %T", v))
+		}
+	}
+
+	s += "{{"
+
+	for i := range v.pattern {
+		if i != 0 {
+			s += " "
+		}
+
+		switch v := v.pattern[i].(type) {
+		case string:
+			s += v
+		case *Expression:
+			s += v.String()
+		default:
+			panic(fmt.Sprintf("unsupported pattern type: %T", v))
+		}
+	}
+
+	return s + "}}"
+}
+
+type local struct {
+	expr     *Expression
+	variable variable
+}
+
+type literal any
+
+type variable string
```

Comment:
I am not really sure about this one, maybe it's there for consistency.

Reply:
in AST.go there is a type variable with an underlying type of string, but it is done to implement interfaces.
Here it is not needed, and it just additional headache because you need to type cast

Reply:
Not sure, why we have `String()` in AST.  Do we have it? AST is for parsing and building AST, and its responsibility is **NOT** to serialize AST into MF2 syntax.

Reply:
AST -> String is a convenient way to convert AST nodes to string, for saving in DB.

**Scenario:**
We want to update message that is stored in our DB
1. We get MF2 string from DB
2. We call `Parse() `to get AST from String
3. We make changes to AST
4. We save AST.String(), into our DB

Reply:
As in our current workflow in translation project, translation works as following:
we extract only translatable text from MF2 AST, translate it, and then replace back into AST.

**So in that Scenario:**
1. We get MF2 string from DB
2. We call `Parse()` to get AST
3. We get only translatable text from AST. (i.e only text nodes)
4. We translate it
5. We replace translatable text nodes with translated text nodes
6. We save `AST.String()`, into DB

Reply:
I'd like to separate all the different concerns. The Builder is for different use cases.

We have 3 responsibilities in the `mf2` repo:
* parsing
* building
* executor (not implemented yet)

If all is good, we'll use AST nodes in the builder's implementation without exposing the implementation.